### PR TITLE
double click to expand or collapse the folder node in tasks view

### PR DESCRIPTION
**Pull request by AndyWu2015**
fix issue:https://github.com/eclipse/buildship/issues/448

### CODE REVIEW

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java [hunk 1]
```diff
@@ -39,10 +42,28 @@ public TreeViewerDoubleClickListener(String commandId, TreeViewer treeViewer) {
...
+        if (provider.hasChildren(selection)) {
+            if (treeViewer.getExpandedState(selection)) {
+                treeViewer.collapseToLevel(selection, AbstractTreeViewer.ALL_LEVELS);
+            }
+            else {
+                treeViewer.expandToLevel(selection, 1);
+            }
+        } else {
+	        NodeSelection nodeSelection = NodeSelection.from(this.treeViewer.getSelection());
+	        if (isEnabledFor(nodeSelection)) {
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java]: The

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java [hunk 2]
```diff
@@ -13,8 +13,11 @@
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java]: The IP validation failed for your pull request:
screen shot 2017-08-07 at 10 19 05

The requirement from the Eclipse Foundation is to [have a `signed-off-by` footer](https://wiki.eclipse.org/Development_Resources/Contributing_via_Git#via_GitHub) in the commit messages.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java [hunk 3]
```diff
@@ -13,8 +13,11 @@
 
 import com.google.common.base.Preconditions;
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java]: This code is missing test coverage. We should have functional tests that
1. imports a project
2. double-clicks on collapsed and expanded project-group-task nodes
3. verifies if the proper expand/collapse/execution action happened.

Please have a look at the `TaskExecutionTest` class, you should be able to create a test based on that one. There are some rough edges there so don't hesitate to ask if something is not clear.

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java [hunk 4]
```diff
@@ -39,10 +42,28 @@ public TreeViewerDoubleClickListener(String commandId, TreeViewer treeViewer) {
 
     @Override
     public void doubleClick(DoubleClickEvent event) {
-        NodeSelection selection = NodeSelection.from(this.treeViewer.getSelection());
-        if (isEnabledFor(selection)) {
-            run();
+        final IStructuredSelection structuredSelection = (IStructuredSelection)event.getSelection();
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java]: I'd use `NodeSelection` here:

```
        NodeSelection nodeSelection = NodeSelection.from(this.treeViewer.getSelection());
        if (nodeSelection.isSingleSelection()) {
            Object selected = nodeSelection.toList().get(0);
            IContentProvider provider = this.treeViewer.getContentProvider();
            if (provider instanceof ITreeContentProvider && ((ITreeContentProvider)provider).hasChildren(selected)) {
                if (this.treeViewer.getExpandedState(selected)) {
                    this.treeViewer.collapseToLevel(selected, AbstractTreeViewer.ALL_LEVELS);
                }
                else {
                    this.treeViewer.expandToLevel(selected, 1);
                }
            }
        }
```

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java [hunk 5]
```diff
@@ -39,10 +42,28 @@ public TreeViewerDoubleClickListener(String commandId, TreeViewer treeViewer) {
...
+            if (treeViewer.getExpandedState(selection)) {
+                treeViewer.collapseToLevel(selection, AbstractTreeViewer.ALL_LEVELS);
+            }
+            else {
+                treeViewer.expandToLevel(selection, 1);
+            }
+        } else {
+	        NodeSelection nodeSelection = NodeSelection.from(this.treeViewer.getSelection());
+	        if (isEnabledFor(nodeSelection)) {
+	            run();
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java]: The task execution should be the default and the node expanding-collapsing should be in the `else` branch.
The current implementation basically masks the feature when the user executes the default project tasks by double-clicking on the project node.

**Comment by AndyWu2015** [comment]
ok , I will do according to the chang request an resend

File: org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java [hunk 6]
```diff
@@ -39,10 +42,28 @@ public TreeViewerDoubleClickListener(String commandId, TreeViewer treeViewer) {
...
+	        }
+	    }
     }
 
     private boolean isEnabledFor(NodeSelection node) {
```
Comment (donat) [inline org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/task/TreeViewerDoubleClickListener.java]: After discussing it internally we concluded that we should disable running the default tasks by double-clicking on the project tasks in favor of expanding-collapsing the node. To implement this, `isEnabledFor(NodeSelection)` should return `TaskViewActionStateRules.taskScopedTaskExecutionActionsEnablement(selection).asBoolean()`.

**Comment by donat** [comment]
@AndyWu2015 I've simplified the task view tests and refactored your test to a [separate class](https://github.com/eclipse/buildship/blob/3b02996c67b706d77a048775d25936ce7be8f59d/org.eclipse.buildship.ui.test/src/main/groovy/org/eclipse/buildship/ui/view/task/TaskViewDoubleClickTest.groovy), please have a look.

**Comment by donat** [comment]
@AndyWu2015 As #347 is a related issue, would you be interested in implementing it?

**Comment by AndyWu2015** [comment]
Thanks for the review and approve , your team are open for contribute , and this is a good for the community , sure , I can do that at my free time.

**Pull request merged**